### PR TITLE
Upgraded library to support most features of WM8960 I2C protocol

### PR DESCRIPTION
**Pull request by relic-se**
I was eager to use SparkFun's WM8960 breakout board with a CircuitPython project, and found this unfinished library to provide a starting point to communicate with this device but otherwise lacking.

To ameliorate this problem, I've ported over the Arduino library and examples provided by SparkFun for this board (https://github.com/sparkfun/SparkFun_WM8960_Arduino_Library) with a few minor modifications:
- `setSampleRate` to automatically configure PLL, system clock, etc for most supported frequencies (8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100, 48000)
- quick configuration methods of key features: `configureI2S`, `configureDAC`, `configureHeadphones`, `configureSpeakers`
- Register reset and general setup on object initialization
- Stereo paired methods for most Left/Right methods to simplify operation
- `setPPLK` uses a single integer value rather than 3 separate bytes

All examples have been tested with an RP2040 and WM8960 breakout module successfully. That is except for those involving I2S input (examples 08, 09, 11, and 15) because I2SIn/I2SInOut is not yet a feature of CircuitPython (as of 9.1.1; more info: https://github.com/adafruit/circuitpython/issues/2676).

I look forward to seeing this library added to the community bundle after review (https://github.com/adafruit/CircuitPython_Community_Bundle) as I think this is a very powerful codec with a large feature set for the price and is already being used in 1 Adafruit product, https://www.adafruit.com/product/4757.

I did my best to give attribution to Pete Lewis of SparkFun, the original author of the Arduino library, under the MIT license, but there may be room for improvement.

Thank you for your consideration!

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
 <!--
 SPDX-FileCopyrightText: 2014 Coraline Ada Ehmke
-SPDX-FileCopyrightText: 2019-2021 Kattni Rembor for Adafruit Industries
+SPDX-FileCopyrightText: 2019-2021 Kattni Rembor
 
 SPDX-License-Identifier: CC-BY-4.0
 -->
-# Adafruit Community Code of Conduct
+# CircuitPython Community Code of Conduct
```
Comment (tannewt) [inline CODE_OF_CONDUCT.md]: Looks like this may have changed due to a community cookiecutter run. Since this is under the adafruit org, then please switch it back.

File: pyproject.toml [hunk 2]
```diff
@@ -12,19 +13,22 @@ requires = [
...
+description = "CircuitPython driver for WM8960 Stereo CODEC"
 version = "0.0.0+auto.0"
 readme = "README.rst"
 authors = [
-    {name = "Adafruit Industries", email = "circuitpython@adafruit.com"}
+    {name = "Adafruit Industries", email = "circuitpython@adafruit.com"},
+    {name = "Cooper Dalrymple", email = "me@dcdalrymple.com"}
 ]
-urls = {Homepage = "https://github.com/adafruit/Adafruit_CircuitPython_WM8960"}
+urls = {Homepage = "https://github.com/dcooperdalrymple/Adafruit_CircuitPython_WM8960"}
```
Comment (tannewt) [inline pyproject.toml]: Undo this change for this repo.

File: adafruit_wm8960.py [hunk 3]
```diff
@@ -21,48 +25,1587 @@
...
-from adafruit_bus_device import i2c_device
-
-import time
+from busio import I2C
+from adafruit_bus_device.i2c_device import I2CDevice
 
 __version__ = "0.0.0+auto.0"
 __repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_WM8960.git"
 
+# I2C address (7-bit format for Wire library)
```
Comment (tannewt) [inline adafruit_wm8960.py]: For any constants below that the user of this library doesn't need, prepend a `_` to the name and wrap the value in `const()`. This will allow mpy-cross to strip the name out of the file and save space.

https://docs.circuitpython.org/en/latest/docs/design_guide.html#use-of-micropython-const

File: adafruit_wm8960.py [hunk 4]
```diff
@@ -21,48 +25,1587 @@
...
+        self._registerLocalCopy = [0x0000 for i in range(len(WM8960_REGISTER_DEFAULTS))]
+        self.reset()
+
+        # General setup
+        self.enableVREF()
+        self.enableVMID()
+
+    def isConnected(self) -> bool:
+        # TODO: Check I2C or I2CDevice
+        return True
```
Comment (tannewt) [inline adafruit_wm8960.py]: I don't think you need this because I2CDevice will check that the device exists on the bus.
```suggestion
```

File: adafruit_wm8960.py [hunk 5]
```diff
@@ -21,48 +25,1587 @@
...
+    # Sets both left and right PGAs
+    def enablePgaZeroCross(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 6, 1) and self._writeRegisterBit(WM8960_REG_RIGHT_INPUT_VOLUME, 6, 1)
+    def disablePgaZeroCross(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 6, 0) and self._writeRegisterBit(WM8960_REG_RIGHT_INPUT_VOLUME, 6, 0)
+
+    def enableLINMUTE(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 1)
+    def disableLINMUTE(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 0) and self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 8, 1)
```
Comment (tannewt) [inline adafruit_wm8960.py]: Instead of have enable and disable functions, please use Adafruit Registers to set bits. https://docs.circuitpython.org/en/latest/docs/design_guide.html#use-adafruit-register-when-possible

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the PR! It seems like this was autoconverted from the Arduino library. That's fine except you now need to curate and refine the API for Python.

Generally we don't want to expose everything. Instead, we only want the APIs we need for the examples we want to show. Your extended examples are a great place to start for this.

I've pointed out a couple things and linked to the dev guide. Please read the dev guide over and refine the API to match that.

Thanks!

**Comment by relic-se** [comment]
Hey Scott! Thanks for looking over this library. I think all of your suggestions so far are on the right track and I will be following your updates shortly.

The only note I'll make is that I did manually convert this library (with extensive use of find and replace, of course), which is why there might be some mistakes here and there. Though I think there is some benefit to maintaining a some compatibility with the original library, I also think it's a great idea to implement the CircuitPython methodology where possible.

However, if a tool does exist to automatically convert Arduino libraries to be compatible with CircuitPython, that would be really great!

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (relic-se) [reply CODE_OF_CONDUCT.md]: I did update some portions of this library based on another recent library. I figured some of the community standards may have changed.

Comment (relic-se) [reply CODE_OF_CONDUCT.md]: Would you like me to roll back the entire code of conduct or just update that one line?

File: adafruit_wm8960.py (hunk 3 above)
Comment (relic-se) [reply adafruit_wm8960.py]: Great idea. What are your thoughts on the prepended `WM8960_...` for user constants? For instance, rather than `adafruit_wm8960.WM8960_MIC_BOOST_GAIN_0DB`, would you prefer `adafruit_wm8960.MIC_BOOST_GAIN_0DB`? The only issue I see is some confusion if the user directly imports the constant, ie `from adafruit_wm8960 import MIC_BOOST_GAIN_0DB`.

File: adafruit_wm8960.py (hunk 5 above)
Comment (relic-se) [reply adafruit_wm8960.py]: I'll work on implementing the Register library. It may take some time as I foresee it dramatically changing the layout of this API. For the better, of course!

File: adafruit_wm8960.py (hunk 3 above)
Comment (relic-se) [reply adafruit_wm8960.py]: Nevermind, I see the convention clearly laid out here. https://docs.circuitpython.org/en/latest/docs/design_guide.html#driver-constant-naming

File: adafruit_wm8960.py (hunk 5 above)
Comment (relic-se) [reply adafruit_wm8960.py]: Actually, we may have an issue here. The WM8960 does not support register reads (see page 62 of the datasheet). I don't think that the adafruit_register library will work for this project in its current state.

From page 62: "If the address is not recognised or **the R/W bit is ‘1’**, the WM8960 returns to the idle condition and wait for a new start condition and valid address."

It may be possible to implement an `adafruit_register.i2c_bit.WOBit` class of which you could declare a default value, add a `reset` method, etc, but that would likely cause unexpected results when using multiple bits of the same register on the same device. A per-device buffer of register contents may be able to fix this but might over complicate the adafruit_register library.

I could make a local class which extends RWBit which fixes some of these issues for this specific device, but I am not sure if that is better than the current implementation.

Comment (tannewt) [reply adafruit_wm8960.py]: The simplest solution may be to use `@property` and cache the last written value.

Comment (tannewt) [reply adafruit_wm8960.py]:
```python
@property
def left_input_mute(self):
    return self._linmute

@left_input_mute.setter(self, value):
    self._linmute = value
    self._write_register_bit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 1 if value else 0)
```

**Comment by tannewt** [comment]
> However, if a tool does exist to automatically convert Arduino libraries to be compatible with CircuitPython, that would be really great!

An LLM like ChatGPT or GitHub Copilot can do this conversion pretty well. Give it the source arduino and ask it to generate python. It may even be able to convert to properties. You'll still need to check and refine it but it makes the initial conversion much quicker.

File: adafruit_wm8960.py (hunk 5 above)
Comment (relic-se) [reply adafruit_wm8960.py]: Properties would be an improvement, and I'll go down that road in case adafruit_register doesn't work out.

Here's an potential example of WOBit I whipped up using a static dictionary of memoryviews. Only the first instance of that register will own the original bytearray object.

This won't work if using different devices which utilize this class with overlapping register addresses, but I can't quite figure out how to pull the appropriate register memoryview without requiring that the owner instance be passed in the constructor.

```python
class WOBit(RWBit):
    _registers = {}

    def __init__(
        self,
        register_address: int,
        bit: int,
        register_width: int = 1,
        lsb_first: bool = True,
        default: bool = False,
    ) -> None:
        RWBit.__init__(self, register_address, bit, register_width, lsb_first)
        if not self.buffer[0] in WOBit._registers:
            WOBit._registers[register_address] = memoryview(self.buffer)
        else:
            del self.buffer
            self.buffer = WOBit._registers[register_address]
        self.default = default
        self.reset()
    
    def reset(self) -> None:
        if self.default:
            self.buffer[self.byte] |= self.bit_mask
        else:
            self.buffer[self.byte] &= ~self.bit_mask

    def __get__(
        self,
        obj: Optional[I2CDeviceDriver],
        objtype: Optional[Type[I2CDeviceDriver]] = None,
    ) -> bool:
        return bool(self.buffer[self.byte] & self.bit_mask)

    def __set__(self, obj: I2CDeviceDriver, value: bool) -> None:
        with obj.i2c_device as i2c:
            if value:
                self.buffer[self.byte] |= self.bit_mask
            else:
                self.buffer[self.byte] &= ~self.bit_mask
            i2c.write(self.buffer)
```

Comment (dcooperdalrymple) [reply adafruit_wm8960.py]: Further issue here found within the datasheet: "The first 7 bits (B15 to B9) are address bits that select which control register is accessed. The remaining 9 bits (B8 to B0) are data bits, corresponding to the 9 bits in each control register." Using the existing RWBit class will push outside the bounds of the register buffer if the 9th bit is accessed. I will either just convert the existing methods to properties or create a dedicated class for register access based on RWBit for this device.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (tannewt) [reply CODE_OF_CONDUCT.md]: I appreciate you updating it. That's fine. I'd just like to keep it the Adafruit version.

File: adafruit_wm8960.py (hunk 5 above)
Comment (tannewt) [reply adafruit_wm8960.py]: You could probably have a shared buffer on `obj` that all of the descriptors mutate. It is weird but works.

Comment (relic-se) [reply adafruit_wm8960.py]: That's a great idea. The reset function would likely need to be called within `__init__` and pass `obj` directly, but that can be done.

Comment (relic-se) [reply adafruit_wm8960.py]: I'm nearly done with my rewrite of the API. There are a large amount of changes and it will likely need to be revised further.

Comment (relic-se) [reply adafruit_wm8960.py]: Here's the latest commit which addresses this request: https://github.com/adafruit/Adafruit_CircuitPython_WM8960/pull/2/commits/e30892ea3797e32911ad559e1522834004ede7ae

I ended up spinning `RWBit` & `RWBits` off as their own classes, `WMBit` & `WMBits`, as the implementation was far too specific for this device.

I am on the fence with some of the property naming conventions:
- Should "...Enabled" be removed where unnecessary?
- What's the best approach to distinguish between Left & Right properties? For instance, should left/right properties be hidden and stereo properties allow for a tuple to specify channel specific settings?
- For properties which offer a unitless direct or a unit-based assignment (ie: "...Volume" vs "...VolumeDb"), should we instead only offer the unit-based property?
- Should "pga" instead be labeled as "mic" since it is primarily used as a the microphone amplifier?

I've yet to update all of the available examples, but most functionality has been ported over with a few extra properties as I've perused the datasheet.

Comment (tannewt) [reply adafruit_wm8960.py]: (Just answering questions. Haven't looked at the changes yet.)

> * Should "...Enabled" be removed where unnecessary?

Ya, when it makes sense. The main thing is that properties are named like the thing rather than featuring a verb.

> * What's the best approach to distinguish between Left & Right properties? For instance, should left/right properties be hidden and stereo properties allow for a tuple to specify channel specific settings?

You could just have `left_` and `right_` variables. Tuple properties will require an allocation but can have the advantage that they get set in the same device write.

> * For properties which offer a unitless direct or a unit-based assignment (ie: "...Volume" vs "...VolumeDb"), should we instead only offer the unit-based property?

Try to match units of other devices so that other things could be swapped in. (I'm not sure if there are other existing drivers.)

> * Should "pga" instead be labeled as "mic" since it is primarily used as a the microphone amplifier?

Not sure. I don't know what "pga" means. Words tend to be better than acronyms.

> I've yet to update all of the available examples, but most functionality has been ported over with a few extra properties as I've perused the datasheet.

Comment (relic-se) [reply adafruit_wm8960.py]: > Ya, when it makes sense. The main thing is that properties are named like the thing rather than featuring a verb.

I'll remove "Enabled" and other verbs where applicable.

> You could just have `left_` and `right_` variables. Tuple properties will require an allocation but can have the advantage that they get set in the same device write.

I'll leave the existing implementation which does this excluding the underscore.

> Try to match units of other devices so that other things could be swapped in. (I'm not sure if there are other existing drivers.)

The only other audio codec library I am aware of is for the VS1053 (https://docs.circuitpython.org/projects/vs1053/en/latest/api.html). I'm not sure if this is the best representation of the policies we want to employ here. Namely, `set_volume(left:int, right:int)` with a range of 0-255. I'd think it would be best to use real-world numbers where possible.

The only downside is when the user would like to control the device on a fully incremental level. Many of the properties of this codec have different decibel minimum and maximums. These would be exposed in API documentation, but the user would have to be aware of the range to utilize the full scope of the property.

> Not sure. I don't know what "pga" means. Words tend to be better than acronyms.

PGA stands for "Programmable-Gain Amplifier". In theory, this input amplifier could be used for a range of devices: electret microphones, high-impedance inputs (ie: a guitar). "mic" designates it as being only usable for 1 thing, which it isn't necessarily the case. Using just "amp" would add confusion with the output mixer and speaker/headphone amps. "inputAmp" may be a viable option, but I'm more inclined to keep it simple with "mic".

Comment (relic-se) [reply adafruit_wm8960.py]: I believe I've tackled most of the points in this conversation and the API is overall on a much better track. Let me know if you have any further comments.

**Comment by relic-se** [comment]
@tannewt What is your take on the examples included here?

I'm planning to remove the examples which require I2S (08, 11, & 15). They both require I2S input. I've tested a PIO-based example for I2S passthrough, but the performance is inadequate without DMA (though it does technically work).

Example #09 requires bluetooth audio communication. I'm not sure if this is possible within CircuitPython, but it seems out of scope. I plan on removing that example as well.

Though I personally like todbot's eighties dystopia, it may also be unnecessary and out of scope.

I think I might just clean house here and only keep what I think is valuable information. The simplification of the API from the original clears things up. I think most use-cases could be expanded upon within the future documentation itself.

File: requirements.txt [hunk 6]
```diff
@@ -3,5 +3,5 @@
 #
 # SPDX-License-Identifier: MIT
 
-Adafruit-Blinka
 adafruit-circuitpython-busdevice
+adafruit-circuitpython-simplemath
```
Comment (tannewt) [inline requirements.txt:7]: Is this for examples only? I'd leave it out if it is.

File: adafruit_wm8960.py [hunk 7]
```diff
@@ -21,48 +25,1164 @@
...
+ALC_MIN_GAIN_MAX = 24.75
+
+GATE_THRESHOLD_MIN = -76.50
+GATE_THRESHOLD_MAX = -30.00
+
+OUTPUT_VOLUME_MIN = -21.00
+OUTPUT_VOLUME_MAX = 0.00
+
+AMP_VOLUME_MIN = -73.00
+AMP_VOLUME_MAX = 6.00
```
Comment (tannewt) [inline adafruit_wm8960.py]: I'd make these private

**Review by tannewt — CHANGES_REQUESTED** [review]
I'd stick to the simpler examples that don't require too much extra. I2SIn examples can always be added later.

The API looks *much* better. Thank you! Could you please add comments to all of the properties now so that the docs show it all? Thanks!

Build instructions are here: https://learn.adafruit.com/creating-and-sharing-a-circuitpython-library/sharing-our-docs-on-readthedocs#sphinx-2615853

File: requirements.txt (hunk 6 above)
Comment (relic-se) [reply requirements.txt:7]: I'm using `adafruit_simplemath.map_range` to scale float decibel values to the integer register values within the library. I could instead use a local private function to do the same task if you'd prefer or a core circuit python function if it exists. The calculations should be very simple, but I didn't think that `adafruit_simplemath` added to much dependency bloat.

File: adafruit_wm8960.py (hunk 7 above)
Comment (relic-se) [reply adafruit_wm8960.py]: I recently switched these back to public to make them accessible if sweeping through the full range of property values. However, these values will be included in the documentation. If you insist, I can revert them back to private.

Also, the `micropython.const` optimization doesn't apply here since it only works on integer values.

**Comment by relic-se** [comment]
I'll begin documentation in the near future, but it might take some time since the functionality of this codec can be convoluted in some aspects.

File: requirements.txt (hunk 6 above)
Comment (tannewt) [reply requirements.txt:7]: Ok, no problem.

**Comment by relic-se** [comment]
I've gotten around to revisiting the examples provided and narrowed them down to only 4 which I felt demonstrated unique features. If I2SInOut ever reaches the light in CircuitPython, I'll make sure to add an example which features ADC to MCU functionality. https://adafruit-circuitpython-wm8960.readthedocs.io/projects/dev/en/latest/examples.html

File: adafruit_wm8960.py [hunk 8]
```diff
@@ -20,49 +24,1617 @@
...
+                value, BOOST_GAIN_MIN, BOOST_GAIN_MAX,
+                1, 7
+            ))
+        )
+        # fmt: on
+
+    @property
+    def input2_boost(self) -> float:
+        value = max(self._left_input2_boost, self._right_input2_boost)
+        # fmt: off
```
Comment (tannewt) [inline adafruit_wm8960.py]: Why disable formatting?

**Review by tannewt — COMMENTED** [review]
Examples look really good. I like how readable the human friendly names are.

One comment about formatting and it looks like you are still working on adding doc strings.

File: adafruit_wm8960.py (hunk 8 above)
Comment (relic-se) [reply adafruit_wm8960.py]: I found that the Black formatting tool would rearrange the map_range parameters in odd ways. I've taken out these tags for now and I'm just letting Black do it's thing, but I'd be lying if I said I was always happy with how it handles formatting in more extreme cases.

**Comment by relic-se** [comment]
@tannewt Here's the updated library. All the documentation has been populated as far as I'm aware. It may get a bit lackadaisical here and there, but I'll leave it up to your scrutiny.

File: adafruit_wm8960/__init__.py [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+
+    LINE3 = 0b100
+    """Connect input 3 as a line input."""
+
+
+# pylint: enable=too-few-public-methods
+
+
+class WM8960:
+    """Driver for controlling the WM8960 audio codec over an I2C connection.Used for receiving line
```
Comment (tannewt) [inline adafruit_wm8960/__init__.py]:
```suggestion
    """Driver for controlling the WM8960 audio codec over an I2C connection. Used for receiving line
```

**Review by tannewt — CHANGES_REQUESTED** [review]
Looks like pre-commit is unhappy. One missing space. Otherwise the comments look really good! This will be a very helpful driver for folks using this chip. Thanks!

File: adafruit_wm8960/__init__.py (hunk 9 above)
Comment (relic-se) [reply adafruit_wm8960/__init__.py]: Yeah, I noticed pre-commit is having issues. I'll run through and fix them this evening. I'm using a Debian machine and definitely struggling to get pre-commit running on my end (I believe this is a known issue related to disttools), otherwise I would have tried to catch that.

**Comment by tannewt** [comment]
I think you need to change this to fix the build issue:
https://github.com/adafruit/Adafruit_CircuitPython_WM8960/blob/cf3620d8ec51535c74e7870ab970a2591f6b8054/pyproject.toml#L41-L44

**Comment by relic-se** [comment]
I think that finally does it! Let me know if there's anything else I need to check up on.

**Review by tannewt — APPROVED** [review]
All good! Thanks for iterating on this with me.

**Comment by relic-se** [comment]
@tannewt Likewise! I'm excited to start using this in my next project.

**Comment by ladyada** [comment]
awesome, thank you @dcooperdalrymple :) 
@TheKitty newsy
